### PR TITLE
#111893 Add Warnings For Missing Index Templates

### LINKED ISSUE #111893: Data stream restoration/promotion without a matching template
Both when we restore a data stream from a snapshot with a new name and when we promote a data stream on the follower cluster of a Cross-Cluster Replication connection, we currently don't check whether the restored/promoted data stream still matches an index template. This doesn't create a problem right away, as users will still be able to write to the data stream's write index, but the data stream will not be able to roll over until a matching template exists. To avoid breaking BWC, we don't want to prevent users from restoring/promoting a data stream when there is no matching index template. Instead, we'll try to make it clear to the user that they're doing something that is likely to cause issues later on. The two ways in which we'd like to make that clear to the user are:
1. Add a warning header to the API request that restored/promoted the data stream
2. Add a warning log

### PR DESCRIPTION
**Pull request by lukewhiting**
Fixes [ES#111893 - Data stream restoration/promotion without a matching template](https://github.com/elastic/elasticsearch/issues/111893)

This PR adds warnings headers and log messages for the following conditions:

* Promoting a CCR replicated data stream wihout a matching index template for the data stream
* Restoring a snapshot of a data stream without a matching index template for the data stream

In both these senarios, the action will succede however future rollovers of the data stream will fail due to the missing templates. By adding a warning header and log, we make the user explicitly aware of this pending failure.

Documentation for both these actions has also been updated to warn of this scenario

### CODE REVIEW

**Comment by lukewhiting** [comment]
es-docs people, Do you need pinging for small changes like this or should I omit the `>docs` label for minor changes?

**Comment by shainaraskas** [comment]
@lukewhiting you don't have to wait for our reviews for small things like these, but keeping the label is a good idea. we don't have the capacity to review every PR, but we try to keep an eye out and provide feedback when possible.

File: x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java
Comment (gmarouli) [inline x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java:1221]: Why are we also trying to clean up the individual backing indices? Is cleaning up the data stream not enough?

**Review by gmarouli — COMMENTED** [review]
Great work @lukewhiting , I learnt a lot reviewing this code. It's touching some parts I am  unfamiliar with. I added some comments and some questions that will help me understand the code better as well.

File: docs/reference/data-streams/promote-data-stream-api.asciidoc
Comment (lukewhiting) [reply docs/reference/data-streams/promote-data-stream-api.asciidoc:23]: Agreed. Wording added.

File: docs/reference/snapshot-restore/index.asciidoc
Comment (lukewhiting) [reply docs/reference/snapshot-restore/index.asciidoc:51]: Agreed. Wording added.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamsSnapshotsIT.java
Comment (lukewhiting) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamsSnapshotsIT.java:1381]: I was trying to ensure there was no chance overlapping templates from the test setup could give a false fail if there was some overlap on the patterns but you're right, in this case it's overkill. Have remove the delete of `with-fs` data stream and the corresponding template.

Though I do still need to delete `other-ds` as it's covered by TEMPLATE_1 and I can't delete the template while a matching DS is still there.

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/PromoteDataStreamTransportAction.java
Comment (lukewhiting) [reply modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/PromoteDataStreamTransportAction.java:116]: To be honest, I didn't know the currentState param was the same thing as came back from the service 😅 But it's super obvious now you pointed it out.

Have switched to using that.

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java
Comment (lukewhiting) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java:524]: I think you are right... There might be a few failed rollover attempts while the restore is in progress (if the DS gets restored before the tamplates) but I believe it would rollover successfully in the end so it's a valid scenario and should not warn.

I'll get that logic in and add some tests for these senario.

File: test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java
Comment (lukewhiting) [reply test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java:919]: That is a really good question and one I have not been able to find the answer to 😅Originally I used a latch with a count of 1 however it caused a deadlock for some reason that hung the test on an internal JVM thread poll method.

I now can't replicate the issue today to might just have been a quirk of my local machine or something broken on `main`. I have switch it to use a latch with a size of 1, lets see if we get any failures in CI.

File: x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java
Comment (lukewhiting) [reply x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java:1139]: Indeed it does :-) Changed

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamsSnapshotsIT.java
Comment (lukewhiting) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamsSnapshotsIT.java:1402]: Yep. Will add it as per https://github.com/elastic/elasticsearch/pull/114589#discussion_r1799340099

File: x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java
Comment (lukewhiting) [reply x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java:1128]: Renamed to be consistent

File: x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java
Comment (lukewhiting) [reply x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java:1152]: The comment reffers to the "step" of these test as a whole rather than that line specifically so perhaps:

```suggestion
            // Prepare index template request and execute
```

File: x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java
Comment (lukewhiting) [reply x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java:1221]: I must admit I copied this from the cleanup step of another test the creates and promotes DataStreams in this class so I'm assuming the original author had their reasons?

I think possibly this ensures the indices are cleaned up, even if the data stream gets itself into into a corrupted state during the test and doing the cleanup prevents other tests in the suite then failing?

File: x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java
Comment (gmarouli) [reply x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java:1152]: It's ok, you can leave it as is. I think my initial confusion is that this request is sent to both clusters but the execution for the leader cluster was a few lines later, but it's not far enough to warrant a more specific comment.

File: x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java
Comment (gmarouli) [reply x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java:1221]: We all have done that :). 

Unfortunately, I also cannot say with confidence that this can never happen. 

I have to admit that I do find this way of "tear down" a bit unusual. I would prefer to have a tear down method that would wipe the cluster at the end of every test so we do not have to deal with this but I do understand that this is out of the scope of this. 

So, happy to keep this, but maybe add a comment, that we do not know if this is necessary but we do it to be sure? So, at least the next maintainer can factor this in. What do you think?

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/PromoteDataStreamTransportAction.java
Comment (gmarouli) [inline modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/PromoteDataStreamTransportAction.java:45]: I think this can also be removed, right?

**Review by gmarouli — APPROVED** [review]
LGTM, I did some minor comments, but apart from these, it's good :) Thank you @lukewhiting 🚀

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/PromoteDataStreamTransportAction.java
Comment (lukewhiting) [reply modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/PromoteDataStreamTransportAction.java]: Yep. Removed :-)

File: x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java
Comment (lukewhiting) [reply x-pack/plugin/ccr/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ccr/AutoFollowIT.java]: Comment added :-)

**Pull request merged**
